src/gluonnlp/data/stream.py

```diff
@@ -213,57 +221,97 @@ def run(self):
         """Method representing the process’s activity."""
         random.seed(self.seed)
         np.random.seed(self.np_seed)
-        mx.random.seed(self.mx_seed)
-
-        stream_iter = iter(self.stream)
+        if not isinstance(self, multiprocessing.Process):
+            # Calling mxnet methods in a subprocess will raise an exception if
+            # mxnet is built with GPU support
+            # https://github.com/apache/incubator-mxnet/issues/4659
+            mx.random.seed(self.mx_seed)
+
+        # Startup - Master waits for this
+        try:
+            stream_iter = iter(self.stream)
+            self._errorq.put(None)
+        except Exception as e:  # pylint: disable=broad-except
+            tb = traceback.format_exc()
+            self._errorq.put((e, tb))
+
+        # Async work
         while True:
             try:  # Check control queue
-                c = self.control_queue.get(False)
+                c = self._controlq.get(False)
                 if c is None:
                     break
+                else:
+                    raise RuntimeError('Got unexpected control code {}'.format(repr(c)))
             except queue.Empty:
                 pass
+            except RuntimeError as e:
+                tb = traceback.format_exc()
+                self._errorq.put((e, tb))
+                self._dataq.put(None)
 
             try:
                 data = next(stream_iter)
-                self.data_queue.put(data)
-            except StopIteration:
-                self.data_queue.put(None)
+                error = None
+            except Exception as e:  # pylint: disable=broad-except
+                tb = traceback.format_exc()
+                error = (e, tb)
+                data = None
+            finally:
+                self._errorq.put(error)
+                self._dataq.put(data)
 
     def __next__(self):
-        next_item = self.data_queue.get()
-        if next_item is None:
-            self.control_queue.put(None)
-            raise StopIteration
-        return next_item
+        next_item = self._dataq.get()
+        next_error = self._errorq.get()
+
+        if next_error is None:
+            return next_item
+        else:
+            self._controlq.put(None)
+            if isinstance(next_error[0], StopIteration):
+                raise StopIteration
+            else:
+                return self._reraise(*next_error)
+
+    def _reraise(self, e, tb):
+        print('Reraising exception from Prefetcher')
```

Comment:
is print appropriate here?

Reply:
The idea is to replicate the standard behaviour of Python printing a traceback before crashing due to an exception. So I guess `print` is appropriate, but let me know if you have another idea. We should `print` to `stderr`. I will push a change for that, thanks for catching this